compiler/rustc_parse/src/lexer/diagnostics.rs

```diff
@@ -0,0 +1,105 @@
+use super::UnmatchedBrace;
+use rustc_ast::token::Delimiter;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_errors::Diagnostic;
+use rustc_span::source_map::SourceMap;
+use rustc_span::Span;
+
+pub fn same_identation_level(sm: &SourceMap, open_sp: Span, close_sp: Span) -> bool {
+    if let (Some(open_padding), Some(close_padding)) =
+        (sm.span_to_margin(open_sp), sm.span_to_margin(close_sp))
+    {
+        open_padding == close_padding
+    } else {
+        false
+    }
+}
+
+// When we get a `)` or `]` for `{`, we should emit help message here
+// it's more friendly compared to report `unmatched error` in later phase
+pub fn report_missing_open_delim(
+    err: &mut Diagnostic,
+    unmatched_braces: &Vec<UnmatchedBrace>,
+) -> bool {
+    let mut reported_missing_open = false;
+    for unmatch_brace in unmatched_braces.iter() {
+        if let Some(delim) = unmatch_brace.found_delim &&
+                matches!(delim, Delimiter::Parenthesis | Delimiter::Bracket) {
+                let missed_open = match delim {
+                    Delimiter::Parenthesis => "(",
+                    Delimiter::Bracket => "[",
+                    _ => unreachable!(),
+                };
+                err.span_label(
+                    unmatch_brace.found_span.shrink_to_lo(),
+                    format!("missing open `{}` for this delimiter", missed_open),
+                );
+                reported_missing_open = true;
+            }
```

Comment:
Nit: formatting is messed up (rustfmt cannot format anything inside `if let` chains).